src/nvme/nbft.c

```diff
@@ -193,14 +200,19 @@ static int read_ssns(struct nbft_info *nbft,
 	__u8 *tmp = NULL;
 	int i, ret;
 
-	if (!(le16_to_cpu(raw_ssns->flags) & NBFT_SSNS_VALID))
+	if (!(le16_to_cpu(raw_ssns->flags) & NBFT_SSNS_VALID)) {
+                errno = EINVAL;
 		return -EINVAL;
+        }
+
 	verify(raw_ssns->structure_id == NBFT_DESC_SSNS,
 	       "invalid ID in SSNS descriptor");
 
 	ssns = calloc(1, sizeof(*ssns));
-	if (!ssns)
+	if (!ssns) {
+                errno = ENOMEM;
```

Comment:
The libnvme convention would be to return -1 after setting errno. Returning the the negative error number *and* setting errno seems strange to me.

Reply:
Does it matter?  Either way the function returns a negative value.  BTW, `-1` is `EPRM`.

```
#define EPERM            1      /* Operation not permitted */
```

Reply:
I hesitate to change this to `-1` because I have no idea if dracut will care.   If you think it's safe to do, I'll go ahead and change everything to -1.  BTW, we have the same issue in nvme-cli.

Reply:
> Does it matter?

@igaw or the other maintainers would need to decide. IMO it matters.

Dracut only distinguishes between zero (success) and non-zero (failure). Shell scripts don't usually rely on any `errno.h` values.

Reply:
The libnvme follows (unfortunatly) the POSIX API design. That means it supposed to  return 1 on error with errno set. We have parts of the API though which returns -1 and sets errno. So my take is that within a subset of the libnvme API (header file) we should aim to be consistent. 

Note, internally we use negative returning negative error codes, this is just for the public API where we
do a return 1, errno=CODE.

Anyway for v2, we have plans to change this part to just return the error codes directly with the option to have more structured error messages means if necessary. Similar to the GLib APIs.

Reply:
Thanks for explaining Daniel.  So all of the internal return -ERROR functions can stay, it's only the public APIs `nbft_free` and `nbft_read` that need to honor the Posix API by using `return 1, errno=ERROR;`  This makes the change much simpler.

Reply:
fixed with my latest update